### PR TITLE
support format

### LINKED ISSUE #62: std::format support
> std::format has its own customization mechanism

### PR DESCRIPTION
**Pull request by alandefreitas**
Fixes #62 

This is the initial proof of concept. I'm still not sure how to organize this into the test since it involves an external dependency.

### CODE REVIEW

File: test/unit/CMakeLists.txt [hunk 1]
```diff
@@ -85,3 +85,20 @@ target_include_directories(boost_url_tests PRIVATE . ../../extra/include)
...
+if (CMAKE_CXX_STANDARD GREATER_EQUAL 14)
+    include(FetchContent)
+    FetchContent_Declare(
+            fmt
+            GIT_REPOSITORY https://github.com/fmtlib/fmt
+            GIT_TAG        8.1.1
+    )
+    FetchContent_MakeAvailable(fmt)
+    target_link_libraries(boost_url_tests PRIVATE fmt::fmt)
+    if (CMAKE_CXX_STANDARD GREATER_EQUAL 17)
```
Comment (alandefreitas) [inline test/unit/CMakeLists.txt:98]: The logic here, for now, was to recur to `__has_include` in C++17 and to a `BOOST_URL_HAS_FMT_FORMAT` macro otherwise.

File: test/unit/CMakeLists.txt [hunk 2]
```diff
@@ -85,3 +85,20 @@ target_include_directories(boost_url_tests PRIVATE . ../../extra/include)
 target_link_libraries(boost_url_tests PRIVATE Boost::url)
 add_test(NAME boost_url_tests COMMAND boost_url_tests)
 add_dependencies(tests boost_url_tests)
+
+if (CMAKE_CXX_STANDARD GREATER_EQUAL 14)
+    include(FetchContent)
+    FetchContent_Declare(
```
Comment (alandefreitas) [inline test/unit/CMakeLists.txt:91]: We have to decide what to do here if we are testing this. An alternative here is simply not testing the integration and testing the formatter directly.

File: test/unit/url.cpp [hunk 3]
```diff
@@ -14,6 +14,55 @@
 #include <algorithm>
 #include <sstream>
 
+#ifdef __has_include
+#if __has_include(<format>)
```
Comment (alandefreitas) [inline test/unit/url.cpp:18]: If we are really implementing, probably move it to config.hpp

File: test/unit/url.cpp [hunk 4]
```diff
@@ -14,6 +14,55 @@
...
+#include <format>
+#endif
+
+#ifdef BOOST_URL_HAS_FMT_FORMAT
+#include <fmt/format.h>
+#endif
+
+#ifdef BOOST_URL_HAS_STD_FORMAT
+template <>
+struct std::formatter<boost::urls::url>
```
Comment (alandefreitas) [inline test/unit/url.cpp:36]: If we are really implementing, probably move it to some kind of formatter header somewhere.

**Comment by vinniefalco** [comment]
Use of `fetchcontent` is questionable. Also, the Visual Studio project generated from this branch does not work, I get these errors:
```
9>The contents of <format> are available only in c++latest mode with concepts support;
9>see https://github.com/microsoft/STL/issues/1814 for details.
9>C:\Users\vinnie\src\boost\libs\url\test\unit\url.cpp(36,13): error C2039: 'formatter': is not a member of 'std'
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\vector(24): message : see declaration of 'std'
9>C:\Users\vinnie\src\boost\libs\url\test\unit\url.cpp(36,22): error C3856: 'formatter': symbol is not a class template
9>C:\Users\vinnie\src\boost\libs\url\test\unit\url.cpp(36,36): error C2143: syntax error: missing ';' before 'boost::urls::url'
9>C:\Users\vinnie\src\boost\libs\url\test\unit\url.cpp(36,1): error C2913: explicit specialization; 'formatter' is not a specialization of a class template
9>C:\Users\vinnie\src\boost\libs\url\test\unit\url.cpp(36,39): error C2059: syntax error: '>'
9>C:\Users\vinnie\src\boost\libs\url\test\unit\url.cpp(37,12): error C2039: 'formatter': is not a member of 'std'
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\vector(24): message : see declaration of 'std'
9>C:\Users\vinnie\src\boost\libs\url\test\unit\url.cpp(37,21): error C2059: syntax error: '<'
9>C:\Users\vinnie\src\boost\libs\url\test\unit\url.cpp(37,27): error C2039: 'string_view': is not a member of 'std'
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\vector(24): message : see declaration of 'std'
9>C:\Users\vinnie\src\boost\libs\url\test\unit\url.cpp(38,1): error C2143: syntax error: missing ';' before '{'
9>C:\Users\vinnie\src\boost\libs\url\test\unit\url.cpp(38,1): error C2447: '{': missing function header (old-style formal list?)
9>C:\Users\vinnie\src\boost\libs\url\test\unit\url.cpp(1868,35): error C2039: 'format': is not a member of 'std'
9>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\vector(24): message : see declaration of 'std'
9>C:\Users\vinnie\src\boost\libs\url\test\unit\url.cpp(1868,1): error C3861: 'format': identifier not found
9>Done building project "boost_url_tests.vcxproj" -- FAILED.
11>------ Build started: Project: ALL_BUILD, Configuration: Debug x64 ------
12>------ Skipped Build: Project: tests, Configuration: Debug x64 ------
12>Project not selected to build for this solution configuration 
13>------ Skipped Build: Project: INSTALL, Configuration: Debug x64 ------
13>Project not selected to build for this solution configuration 
========== Build: 5 succeeded, 1 failed, 0 up-to-date, 7 skipped ==========
```

I think we should probably only support `<format>`, conditionally compiled for C++20 (?) and maybe not worry about the standalone fmt library for now.

**Comment by alandefreitas** [comment]
Do we need a function to return a `string_view` from a `url_view`?

`std::string_view(u.c_str(), u.size())` looks ugly.

**Comment by vinniefalco** [comment]
> Do we need a function to return a string_view from a url_view?

What about `url_view::encoded_url()` ?

File: include/boost/url/url.hpp [hunk 5]
```diff
@@ -1413,6 +1416,13 @@ operator<<(std::ostream& os, url const& u);
 } // urls
 } // boost
 
+#ifdef BOOST_URL_HAS_STD_FORMAT
+/** Format the encoded url with std::format
+*/
+template <>
+struct std::formatter<boost::urls::url>;
```
Comment (vinniefalco) [inline include/boost/url/url.hpp:1423]: Use `::boost::urls::url`

Comment (vinniefalco) [inline include/boost/url/url.hpp:1423]: I don't think this should be part of the reference

File: include/boost/url/impl/url.ipp [hunk 6]
```diff
@@ -2343,4 +2343,20 @@ operator<<(
 } // urls
 } // boost
 
+#ifdef BOOST_URL_HAS_STD_FORMAT
+template <>
+struct std::formatter<boost::urls::url>
+    : std::formatter<std::string_view>
```
Comment (vinniefalco) [inline include/boost/url/impl/url.ipp:2349]: You will have to include `<string_view>` for this, no?

File: include/boost/url/impl/url.ipp [hunk 7]
```diff
@@ -2343,4 +2343,20 @@ operator<<(
...
+        const boost::urls::url& u,
+        FormatContext& ctx)
+    {
+        return formatter<string_view>::format(
+            std::string_view(u.c_str(), u.size()), ctx);
+    }
+};
 #endif
+
+#endif
```
Comment (vinniefalco) [inline include/boost/url/impl/url.ipp:2362]: missing newline

File: include/boost/url/impl/url.ipp [hunk 8]
```diff
@@ -2343,4 +2343,20 @@ operator<<(
...
+struct std::formatter<boost::urls::url>
+    : std::formatter<std::string_view>
+{
+    template <typename FormatContext>
+    auto format(
+        const boost::urls::url& u,
+        FormatContext& ctx)
+    {
+        return formatter<string_view>::format(
+            std::string_view(u.c_str(), u.size()), ctx);
```
Comment (vinniefalco) [inline include/boost/url/impl/url.ipp:2357]: `u.string()`

Comment (vinniefalco) [reply include/boost/url/impl/url.ipp:2357]: Also you should use `u.data()` instead of `u.c_str()` (if you wanted the pointer and size)

**Comment by alandefreitas** [comment]
> What about url_view::string() ?

At first, I assumed it would return a `std::string`. But then I realized ensuring std::string_view would be better anyway.

**Comment by alandefreitas** [comment]
I've been wondering we should really keep pursuing this PR. It's applicability seems very limited:

- For a given `url` or `url_view`, the user can always `fmt::format("{}", u.string())`. The formatter is just a long way to replicate this logic.
- The `fmt` is the only implementation people seem to be using in practice. It's also the best implementation. The `fmt` formatter worked perfectly in tests, but integrating a non-boost external dependency just for that is controversial, to say the least. 
- The formatter is just a customization point, so we could just leave the formatter there. But then there would be no integration tests.
- GCC and Clang don't implement `std::format` yet, so MSVC code using `std::format` will not work everywhere.
- The MSVC implementation has a bug where the `string_view` formatter is not `const`, so the usual strategy of inheriting from `string_view` doesn't work. This would require us to reimplement the whole `string_view` formatter, which is probably the most complex formatter, since it handles all formatting options available in `fmt`.

**Comment by vinniefalco** [comment]
tl;dr: too much try-hard for too little reward
